### ISSUE TITLE
Localization and internationalization

### ISSUE THREAD

**Issue by goanpeca** [napari/napari#2195]
## 🚀 Feature

Add support for working with Napari in a different UI display language. e.g. Klingon.

## Motivation

Whenever a piece of software becomes widely used, more and more users want to use it. Some of these users might benefit from having the user interface, be displayed in their native language, different from English.

## Pitch

Democratizing open source goes beyond actual source code :), it should alleviate language barriers whenever possible!

## Alternatives

N.A.

## Additional context

Generating translations requires doing some i18n first, so I suggest creating a `translation.py` module, in `napari/utils` or similar and create a dummy `_` method using the `gettext.gettext` convention (the standard with Python) and replace all strings from `"Some string"` to  `_("Some string")`, that way we can start enforcing this throughout the codebase.

```python
# napari/utils/translation.py
def _(string):
    """Dummy gettext.gettext equivalent to prepare strings for localization."""
    return string

    
# napari/...
from napari.utils.translation import _

a = _("Something something")
```

Additional to this, by doing this change, it will allow us to think a bit harder on possible cases where strings could be defined in a different way to be more clear.

**Comment by Czaki** [comment]
It should be well documented and put in `napari_plugin_engine` instead of `napari` package. Internationalization without support to plugins (for widgets) is not a big profit (because of language mixes in UI)

**Comment by sofroniewn** [comment]
> It should be well documented and put in napari_plugin_engine instead of napari package. Internationalization without support to plugins (for widgets) is not a big profit (because of language mixes in UI)

I like this idea, but we can also get started with something in `utils` first just to get something off the ground. I wonder if translation is easier for widgets we autogenerate (like with magicgui) vs custom widgets - could be another benefit of autogeneration!

**Comment by Czaki** [comment]
> I like this idea, but we can also get started with something in `utils` first just to get something off the ground. I wonder if translation is easier for widgets we autogenerate (like with magicgui) vs custom widgets - could be another benefit of autogeneration!

I still to add dummy function to `napari-plugin-engine` as fast as possible and update documentation. Even if we start implementation in `napari.utils` then part of the plugins will be ready for translation. 

Alongside the `_` function there should be a hook to provide plugin side fallback translation.

**Comment by goanpeca** [comment]
Thanks for the pointers @Czaki.

Just to add some more information to take into account moving forward. It is better to have the concept of a language pack that aggregates the translations for napari (core), and any plugin that wants to provide them, that way we centralize all the efforts in a single repo and a single place (with a bunch of nice automations)

This is an example using crowdin which was setup for JLab:
- http://crowdin.com/projects/jupyterlab
- https://github.com/jupyterlab/language-packs

**Comment by sofroniewn** [comment]
This is so cool! I wasn't familiar with crowdin before
![image](https://user-images.githubusercontent.com/6531703/106494564-701c9180-646f-11eb-852a-0a8faf3e100d.png)

**Comment by goanpeca** [comment]
Yeah it is pretty cool :), there are too many services to pick from, but we have used this for several projects, and they have been very responsive and helpful.

**Comment by goanpeca** [comment]
@Czaki could you elaborate a bit more on:

>Alongside the _ function there should be a hook to provide plugin side fallback translation.

**Comment by Czaki** [comment]
For example, I could write a plugin that provides some measurements method to napari. This measurement could have names and descriptions which are not already presented in the napari translation file. It will be nice If I could provide such translations for languages that I know inside the plugin, not creating PR to some main package.

**Comment by goanpeca** [comment]
Hi @Czaki so maybe it is worth explaining how localization will (or at least in my head) will work

Any plugin developer will do something like

```python

from napari.utils.translations import translator
# Or when move to plugin engine
from napari_plugin_engine.utils.translations import translator   # or similar.

trans = translator.load("napari_console")  # Using the napari console example

# hat provides some measurements method to napari
# And then you would use this anywhere you need to provide translations

something = trans._("something")
something_else = trans._("something else")
```

This is what devs will need to do. From there we will centralize the generation of catalog for any registered plugins in the language packs repository.

And example of the configuration file on that repo from another project where I followed a similar approach
<img width="476" alt="Screen Shot 2021-02-19 at 12 08 35 PM" src="https://user-images.githubusercontent.com/3627835/108537095-32ab6700-72ab-11eb-9446-129d7d70891e.png">


From that configuration file we will automatically scrape for strings ("something" and "something_else") and save it  in a catlog named `napari_console.pot` and that will also connect to crowdin service or similar.

A language-pack will contain the compiled catalgos (`*.mo`) for each separate thing so `napari.mo`, `napari_console.mo` etc.

**Comment by Czaki** [comment]
Maybe I do not understand something. But how it allows for the translation of a text which is not already present in napari translation files (for example during the development of a new plugin?.

**Comment by goanpeca** [comment]
>But how it allows for the translation of a text which is not already present in napari translation files (for example during the development of a new plugin?.

Until the code is using the `translator` and the strings have been scrapped and they have been uploaded to crowdin and a language pack has been generated, then no you cannot do it.

However for a plugin in development, as long as there is a public repo, you can add the things in here

![image](https://user-images.githubusercontent.com/3627835/121105577-f8b69000-c7c9-11eb-9964-04d4103d9b8e.png)

And point to master (not a tag) then you could locally install the language pack and tests things.
